crates/bevy_ecs/src/bundle.rs

```diff
@@ -166,19 +266,277 @@ impl BundleInfo {
         });
     }
 
+    /// Adds a bundle to the given archetype and returns the resulting archetype. This could be the same
+    /// [ArchetypeId], in the event that adding the given bundle does not result in an Archetype change.
+    /// Results are cached in the Archetype Graph to avoid redundant work.
+    pub(crate) fn add_bundle_to_archetype(
+        &self,
+        archetypes: &mut Archetypes,
+        storages: &mut Storages,
+        components: &mut Components,
+        archetype_id: ArchetypeId,
+    ) -> ArchetypeId {
+        if let Some(add_bundle) = archetypes[archetype_id].edges().get_add_bundle(self.id) {
+            return add_bundle.archetype_id;
+        }
+        let mut new_table_components = Vec::new();
+        let mut new_sparse_set_components = Vec::new();
+        let mut bundle_status = Vec::with_capacity(self.component_ids.len());
+
+        let current_archetype = &mut archetypes[archetype_id];
+        for component_id in self.component_ids.iter().cloned() {
+            if current_archetype.contains(component_id) {
+                bundle_status.push(ComponentStatus::Mutated);
+            } else {
+                bundle_status.push(ComponentStatus::Added);
+                // SAFE: component_id exists
+                let component_info = unsafe { components.get_info_unchecked(component_id) };
+                match component_info.storage_type() {
+                    StorageType::Table => new_table_components.push(component_id),
+                    StorageType::SparseSet => {
+                        storages.sparse_sets.get_or_insert(component_info);
+                        new_sparse_set_components.push(component_id)
+                    }
+                }
+            }
+        }
+
+        if new_table_components.is_empty() && new_sparse_set_components.is_empty() {
+            let edges = current_archetype.edges_mut();
+            // the archetype does not change when we add this bundle
+            edges.set_add_bundle(self.id, archetype_id, bundle_status);
```

Comment:
This wasn't added in this PR, but I would prefer a rename of this method (or at least some compelling doc strings). The functionality of this line is very unclear.

Reply:
Yeah this this is an `insert` operation (not `set`). But changing that doesn't really help with the unfortunate `(insert or set) | add` adjacency (which is what I think your comment is largely about). I think `insert_X` where `X` is `AddEdge` is still idiomatic. Its just _also_ dumb :smile: Idk how to reconcile the two without obscuring either the operation being done or the type being inserted.  

I'm also thinking about breaking out edges into their own `HashMap<(ArchetypeId, BundleId), AddEdge>` data structure for a couple of reasons:

1. The current sparse set approach is pretty memory hungry
2. Archetypes owning edges creates some unfortunate ownership issues when trying to batch operations. We need to re-lookup the edge on every insert instead of caching it once at the beginning.

We could re-visit AddEdge naming / structure there too. But that should all be in a separate pr. For now I think more docs (and renaming set to insert) is the way to go.

Reply:
Insert_edge_addition perhaps?

Reply:
> I'm also thinking about breaking out edges into their own HashMap<(ArchetypeId, BundleId), AddEdge> data structure

This seems reasonable.

> We could re-visit AddEdge naming / structure there too. But that should all be in a separate pr. For now I think more docs (and renaming set to insert) is the way to go.

Sounds good to me.